packages/pds/src/sql-repo-storage.ts

```diff
@@ -0,0 +1,200 @@
+      .select('content')
+      .executeTakeFirst()
+    return found ? found.content : null
+  }
+
+  async hasSavedBlock(cid: CID): Promise<boolean> {
+    const found = await this.db.db
+      .selectFrom('ipld_block')
+      .where('cid', '=', cid.toString())
+      .select('cid')
+      .executeTakeFirst()
+    return !!found
+  }
+
+  async putBlock(cid: CID, block: Uint8Array): Promise<void> {
+    this.db.assertTransaction()
+    const insertBlock = this.db.db
+      .insertInto('ipld_block')
+      .values({
+        cid: cid.toString(),
+        size: block.length,
+        content: block,
+        indexedAt: this.timestamp || new Date().toISOString(),
+      })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    const insertCreator = this.db.db
+      .insertInto('ipld_block_creator')
+      .values({
+        cid: cid.toString(),
+        did: this.did,
+      })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    await Promise.all([insertBlock, insertCreator])
+  }
+
+  async putMany(toPut: Map<string, Uint8Array>): Promise<void> {
+    this.db.assertTransaction()
+    const blocks: IpldBlock[] = []
+    const creators: IpldBlockCreator[] = []
+    for (const staged of toPut.entries()) {
+      const [cid, bytes] = staged
+      blocks.push({
+        cid: cid.toString(),
+        size: bytes.length,
+        content: bytes,
+        indexedAt: this.timestamp || new Date().toISOString(),
+      })
+      creators.push({
+        cid: cid.toString(),
+        did: this.did,
+      })
+    }
+    const insertBlocks = this.db.db
+      .insertInto('ipld_block')
+      .values(blocks)
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    const insertCreators = this.db.db
+      .insertInto('ipld_block_creator')
+      .values(creators)
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    await Promise.all([insertBlocks, insertCreators])
+  }
+
+  async commitStaged(commit: CID, prev: CID | null): Promise<void> {
+    this.db.assertTransaction()
+    const commitBlocks: RepoCommitBlock[] = []
+    this.staged.forEach((_bytes, cid) => {
+      commitBlocks.push({
+        commit: commit.toString(),
+        block: cid.toString(),
+      })
+    })
+    const insertBlocks = this.putMany(this.staged)
+    const insertCommit = this.db.db
+      .insertInto('repo_commit_block')
+      .values(commitBlocks)
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    const updateRoot =
+      prev === null ? this.insertRoot(commit) : this.updateRoot(commit, prev)
+    const insertCommitHistory = this.db.db
+      .insertInto('repo_commit_history')
+      .values({
+        commit: commit.toString(),
+        prev: prev ? prev.toString() : null,
+      })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    await Promise.all([
+      insertBlocks,
+      insertCommit,
+      updateRoot,
+      insertCommitHistory,
+    ])
+    this.clearStaged()
```

Comment:
I assume it's cool for the blocks to remain staged even if this promise rejects, just flagging for a second look.

Reply:
yeah i think you'd want them too, because you may want to retry
that being said, my next PR takes a diff approach & removes staged blocks entirely